### PR TITLE
Feat/grouping statistics

### PR DESCRIPTION
**Pull request by gofmanaa**
added group statistics by source and daemonIds before send

### CODE REVIEW

File: mamoru-sniffer/src/sniffer.rs [hunk 1]
```diff
@@ -54,6 +54,9 @@ pub struct SnifferConfig {
 
     #[serde(default = "SnifferConfig::default_statistics_buffer_size")]
     pub statistics_buffer_size: usize,
+
+    #[serde(default = "SnifferConfig::default_statistics_send_interval_memool_milsec")]
+    pub statistics_send_interval_memool_milsec: u64,
```
Comment (sokyrko) [inline mamoru-sniffer/src/sniffer.rs:59]: default_statistics_send_interval_mempool_ms

File: mamoru-sniffer/src/statistics_bg_task.rs [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
-use tracing::{debug, error};
+use tracing::{debug, error, info};
 
 #[derive(Debug, Copy, Clone, Deserialize)]
 pub struct BgStatisticsConfig {
     #[serde(default = "BgStatisticsConfig::default_send_interval_sec")]
     pub send_interval_sec: Option<u64>,
 
+    #[serde(default = "BgStatisticsConfig::default_send_interval_mempool_millis")]
+    pub send_interval_mempool_millis: u64,
```
Comment (sokyrko) [inline mamoru-sniffer/src/statistics_bg_task.rs:18]: _ms

File: chain-client/src/config.rs [hunk 3]
```diff
@@ -154,7 +154,7 @@ impl ChainConfig {
     }
 
     fn default_tx_gas_limit() -> u64 {
-        300_000
+        400_000
```
Comment (sokyrko) [inline chain-client/src/config.rs:157]: What is the reason for this change?

File: mamoru-sniffer/src/sniffer.rs [hunk 4]
```diff
@@ -49,11 +49,14 @@ pub struct SnifferConfig {
...
-    #[serde(default = "SnifferConfig::default_statistics_send_interval_secs")]
-    pub statistics_send_interval_secs: Option<u64>,
+    #[serde(default = "SnifferConfig::default_statistics_send_interval_millis")]
+    pub statistics_send_interval_millis: u64,
 
     #[serde(default = "SnifferConfig::default_statistics_buffer_size")]
     pub statistics_buffer_size: usize,
+
+    #[serde(default = "SnifferConfig::default_statistics_send_interval_memool_millis")]
+    pub statistics_send_interval_memool_millis: u64,
```
Comment (sokyrko) [inline mamoru-sniffer/src/sniffer.rs:59]: `memool` -> `mempool`

File: mamoru-sniffer/src/statistics_bg_task.rs [hunk 5]
```diff
@@ -1,112 +1,146 @@
...
+        } else {
+            i += 1;
         }
+    }
+    result
+}
 
-        Ok(items)
+/// Receive statistic from sniffer and send it to Validation Chain.
+pub async fn receive_statistic(
```
Comment (sokyrko) [inline mamoru-sniffer/src/statistics_bg_task.rs:92]: Let's update the doc block and make name to be like `receive_statistics_job`

File: mamoru-sniffer/src/statistics_bg_task.rs [hunk 6]
```diff
@@ -1,112 +1,146 @@
...
+    while let Some(statistics) = statistic_rx.recv().await {
+        let mut accumulator = accumulator.lock().await;
+        let mut new_statistics = vec![statistics];
+        new_statistics.extend(accumulator.clone());
+        *accumulator = group_statistics(new_statistics);
     }
 }
+
+/// Group statistics by source and daemons_ids
+fn group_statistics(items: Vec<StatisticsReport>) -> Vec<StatisticsReport> {
```
Comment (sokyrko) [inline mamoru-sniffer/src/statistics_bg_task.rs:105]: Where are `group_statistics`, `take_by_source` tests?

File: mamoru-sniffer/src/statistics_bg_task.rs [hunk 7]
```diff
@@ -1,112 +1,146 @@
...
+    // Group items by source and daemons_ids and calculate the maximum values
+    let mut grouped: HashMap<(SourceType, Vec<String>), StatisticsReport> = HashMap::new();
```
Comment (sokyrko) [inline mamoru-sniffer/src/statistics_bg_task.rs:107]: I think the key must include `HashSet`, not `Vec` of daemon ids, as the order may be different.

File: mamoru-sniffer/tests/validation_chain_tests/statistics.rs [hunk 8]
```diff
@@ -9,64 +9,7 @@ use mamoru_sniffer::client::{
...
-
-        Ok(())
-    })
-    .await
-    .expect("Failed to query statistics.");
-}
+use crate::validation_chain_tests::{retry, sniffer, sniffer::random_string};
 
 #[test(tokio::test)]
 #[ignore]
```
Comment (sokyrko) [inline mamoru-sniffer/tests/validation_chain_tests/statistics.rs:15]: Please update the smoke test to check that the statistics were grouped.

File: mamoru-sniffer/src/statistics_bg_task.rs (hunk 7 above)
Comment (gofmanaa) [reply mamoru-sniffer/src/statistics_bg_task.rs:107]: Unable to use HashSet in HashMap
error[E0599]: the method `entry` exists for struct `HashMap<(SourceType, HashSet<String>), StatisticsReport>`, but its trait bounds were not satisfied

**Review by sokyrko — APPROVED** [review]
